### ISSUE TITLE
[inductor] int64 scatter_ (reduce=None) on GPU devices falls back without deterministic algorithms, unlike index_put_

### ISSUE THREAD

**Issue by spsicio** [pytorch/pytorch#194187]
### 🐛 Describe the bug

Without deterministic algorithms enabled, `scatter_` with `reduce=None` is lowered to a native fallback when `src` has dtype `torch.int64` on GPU devices, whereas the semantically equivalent `index_put_` with `accumulate=False` does not fall back and emits a Triton kernel. This inconsistency is surprising, and the fallback for `scatter_` is unnecessary: `reduce=None` performs a plain store, which does not require the atomic-add path that the int64 fallback is meant to guard against.

```python
import torch

torch.manual_seed(0)
dev = "cuda"


def fn(x, idx, src):
    # x.index_put_((idx,), src)
    x.scatter_(0, idx, src)
    return x


torch._inductor.config.fx_graph_cache = False

x = torch.zeros(64, dtype=torch.int64, device=dev)
idx = torch.randperm(64, device=dev)[:32]
src = torch.randint(1, 9, (32,), dtype=torch.int64, device=dev)

torch.compile(fn)(x, idx, src)
```

Run and inspect the lowering:

```bash
export TORCH_LOGS="+torch._inductor.scheduler"
python script.py 2>&1 | grep ScatterFallback
```

With `x.scatter_(0, idx, src)` uncommented, the log contains ScatterFallback. With `x.index_put_((idx,), src)` uncommented instead, ScatterFallback does not appear.

And why `scatter_` falls back? In `torch/_inductor/lowering.py`, `scatter_` decides whether to fall back via `use_scatter_fallback` in `torch/_inductor/utils.py`. The relevant condition is:

```python
src_is_tensor
and is_gpu(src_device_type)
and needs_fallback_due_to_atomic_add_limitations(src_dtype)
```

There is no check on `reduction_type` here, so `reduce=None` also triggers the fallback.

### Versions

```
PyTorch version: 2.12.1+cu130
CUDA used to build PyTorch: 13.0
CUDA runtime version: 13.2.78
GPU models and configuration: GPU 0: NVIDIA RTX PRO 5000 72GB Blackwell
Nvidia driver version: 595.80
Python version: 3.10.20

Versions of relevant libraries:
[pip3] torch==2.12.1
[pip3] triton==3.7.1
```

cc @pragupta @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by TANGBUDU** [comment]
I reproduced this locally on an RTX 5070 Ti with PyTorch 2.12.1+cu130 and reviewed the relevant Inductor fallback path. I’d like to take responsibility for preparing and validating a focused fix.

> The following technical summary was drafted with ChatGPT assistance:
>
> `TORCH_LOGS=+torch._inductor.scheduler` shows `ScatterFallback` for
> `int64 scatter_(reduce=None)`, while the equivalent
> `index_put_(accumulate=False)` path does not fall back.
>
> The intended change is to avoid that unnecessary fallback for the
> non-reduction case while preserving the current fallback behavior for
> int64 reductions and deterministic mode.

If this is suitable for an external contribution, could a maintainer mark the issue `actionable`? I would then prepare a small fix and a regression test for the non-reduction case that fails on unmodified `main`, and validate the existing reduction and deterministic behavior separately.

**Comment by spsicio** [comment]
> I reproduced this locally on an RTX 5070 Ti with PyTorch 2.12.1+cu130 and reviewed the relevant Inductor fallback path. I’d like to take responsibility for preparing and validating a focused fix.
> 
> > The following technical summary was drafted with ChatGPT assistance:
> > `TORCH_LOGS=+torch._inductor.scheduler` shows `ScatterFallback` for
> > `int64 scatter_(reduce=None)`, while the equivalent
> > `index_put_(accumulate=False)` path does not fall back.
> > The intended change is to avoid that unnecessary fallback for the
> > non-reduction case while preserving the current fallback behavior for
> > int64 reductions and deterministic mode.
> 
> If this is suitable for an external contribution, could a maintainer mark the issue `actionable`? I would then prepare a small fix and a regression test for the non-reduction case that fails on unmodified `main`, and validate the existing reduction and deterministic behavior separately.

Thanks for being willing to work on this fix!

If it's okay with you, would you mind adding me as a contributor in the commit, using either `Co-authored-by: spsicio <spsicio@foxmail.com>` or `Reported-by: spsicio <spsicio@foxmail.com>`? I would be very happy to review and test your code as well.

**Comment by TANGBUDU** [comment]
No problem at all. I’ll use Reported-by: spsicio <spsicio@foxmail.com> in the commit. Thanks again for reporting the issue and for offering to review and test the fix.
